calicoctl/commands/ipam.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2016 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/tigera/libcalico-go/lib/client"
+	caliconet "github.com/tigera/libcalico-go/lib/net"
+
+	docopt "github.com/docopt/docopt-go"
+)
+
+// IPAM show/release is for managing Calico IPAM by showing the
+// current use of an IP and/or releasing the IP address from use
+func IPAM(args []string) error {
+	doc := `Usage: 
+    calicoctl ipam release <IP>
+    calicoctl ipam show <IP>
+
+Description:
+    Manage Calico assigned IP addresses
+
+Warnings:
+  -  Releasing an in-use IP address can result in it being assigned to multiple
+     workloads.
+`
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	// length of parsedArgs is 0 when `-h` `--help` is passed
+	// docopts takes care of printing the help message, but we need to
+	// stop return nil here so command doesn't go any further when `-h` is passed
+	if len(parsedArgs) == 0 {
+		return nil
+	}
+
+	// Parse the input string as an IP address (IPv4 and IPv6)
+	// This also validates the IP address
+	ip := net.ParseIP(args[2])
+	if ip == nil {
+		fmt.Println("Invalid IP address specified.")
+		os.Exit(1)
+	}
+
+	// create a new backend client from env vars
+	backendClient, err := newClient("")
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	ipamClient := backendClient.IPAM()
+
+	switch args[1] {
+	case "show":
+		err = showIP(ipamClient, ip)
+	case "release":
+		err = releaseIP(ipamClient, ip)
+	}
+
+	// Exit with an error message if any of the command returned an error
+	if err != nil {
+		fmt.Printf("Error executing command %v, Error: %v\n", args, err)
+		os.Exit(1)
+	}
+
+	return nil
+}
+
+// showIP gets the attributes of an IP address, and returns nil if it is assigned
+// or an error with a message if not assigned
+func showIP(ipamClient client.IPAMInterface, ip net.IP) error {
+
+	attr, err := ipamClient.GetAssignmentAttributes(caliconet.IP{ip})
+
+	// IP address is assigned
+	if err == nil {
+
+		// IP address is assigned with attributes
+		if len(attr) != 0 {
+			fmt.Println(attr)
+		} else {
+			// IP address is assigned but attributes are not set
+			fmt.Printf("No attributes defined for IP %s\n", ip)
+		}
+	} else {
+		// IP address is not assigned, this prints message like
+		// `IP 192.168.71.1 is not assigned in block`
+		fmt.Println(err)
```

Comment:
I've restructured the whole conditional logic, but the error can't be returned to the caller because it is not exactly an error (it's more of a message to print. I just added a comment on top to explain it), and in Go, idiomatic way to handle error is to handle them where they happen instead of bubbling them up (we don't have that in our code base everywhere but I'd like to follow that best practice for the new code that gets written)